Replace org.apache.hadoop.fs dependency in DynamicSubmitter with com.google.cloud.storage

krisss85 opened the pull request.

Xeago commented on common/src/main/java/com/spotify/spydra/util/StorageUtils.java:
```diff
@@ -0,0 +1,82 @@
+
+
+public class StorageUtils {
+    private static final Storage storage = StorageOptions.getDefaultInstance().getService();
```
I suppose this internally uses ApplicationDefaultCredentials?

krisss85 replied:
Yes indeed. Will reuse what is done in GCPUtils. Maybe I should even move all these methods there?

Xeago commented on common/src/main/java/com/spotify/spydra/util/StorageUtils.java:
```diff
@@ -0,0 +1,82 @@
+
+    public Bucket checkBucket(String bucketName) {
+        if (bucketName == null) {
+            throw new IllegalArgumentException(
```
I think I prefer the pattern of [`Objects.requireNonNull(bucketName, 'Provide bucket name.')`](https://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#requireNonNull-T-java.lang.String-). Though this is fine too.

Xeago commented on spydra/src/main/java/com/spotify/spydra/submitter/api/DynamicSubmitter.java:
```diff
@@ -213,9 +212,13 @@ private void waitForHistoryToBeMoved(SpydraArgument arguments) throws IOExceptio
+    }
     LOGGER.info("Waiting for history files to be moved to its final location");
-    while (fs.getContentSummary(new Path(uri)).getFileCount() != 0) {
+    while (new StorageUtils().getFileCount(bucketName, directory+"/") != 0) {
```
This constructs a new StorageUtils on every iteration.

Xeago replied:
It also pages through the entire object listing on GCS, I don't think there's a concern here though.
Maybe if it's a pooling cluster (thus lots of objects), and the cluster breaks we might spend a lot of time iterating here.

krisss85 replied:
Didn't find a method to quickly check the size of the directory. Maybe I should just implement a methos isEmpty() and stop is hasNext() returns anything?

Xeago replied:
I think the abstraction of `StorageUtil` here creates more confusion and conflict between organising code and implementing the logic we would like.

varjoranta commented on spydra/src/main/java/com/spotify/spydra/submitter/api/DynamicSubmitter.java:
```diff
@@ -213,9 +212,13 @@ private void waitForHistoryToBeMoved(SpydraArgument arguments) throws IOExceptio
+    String bucketName = uri.getHost();
+    String directory = uri.getPath();
+    if (directory != null && directory.length() > 0) {
+      directory = directory.substring(1, directory.length()); //remove trailing slash from path
```
This actually removes the leading slash from the path.

krisss85 replied:
ahhh yes

krisss85 commented:
@Xeago @varjoranta 
Thanks for the comments so far. Do you think it is worth to continue with the general approach?

varjoranta commented on common/src/main/java/com/spotify/spydra/util/StorageUtils.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    public int getFileCount(String bucketName, String directory) {
+        Iterator bucketIterator = listBucket(bucketName, directory).getValues().iterator();
```
Notice that you are counting here the objects on first page, not overall. Compare iterateAll() with getValues()

varjoranta replied:
I also remember you wondering yourself whether we should do full object count just to decide whether target exists (!= 0). I think we should improve here, but don't know top of my head how.

varjoranta commented on spydra/src/main/java/com/spotify/spydra/submitter/api/DynamicSubmitter.java:
```diff
@@ -213,9 +212,13 @@ private void waitForHistoryToBeMoved(SpydraArgument arguments) throws IOExceptio
+    }
     LOGGER.info("Waiting for history files to be moved to its final location");
-    while (fs.getContentSummary(new Path(uri)).getFileCount() != 0) {
+    while (new StorageUtils().getFileCount(bucketName, directory+"/") != 0) {
```
I would use something else than file count here, just to avoid the lazy paging happening in the getFileCount method.

varjoranta commented:
In general I'm fine with this approach. Just to get the small things fixed and then not to do full paging just to check count != 0. Usage of StorageUtils is also something we might not need, but that's more about structure, not about criticising the general implementation.

krisss85 commented:
@xeago @varjoranta I updated the PR. Since there are some substantial changes you may want to do another full review. I have moved the GCS operations to GcpUtils. Regarding the count I used java8 stream and spliterator. There are some ways to do this with apache.common or guava, but wanted to stay with pure java. Let me know what do you think of these changes.
